### ISSUE TITLE
Schema defer directive: Do not add it to a supergraph schema if it's already present

### ISSUE THREAD

**Issue by o0Ignition0o** [apollographql/router#1962]
Spinning up the router with a schema that already contains the `@defer` directive definition currently prevents the router from starting.

Unfortunately we don't have any control over the subgraph schema generation, and some tools will append the directive automatically.

```
cannot create the router: couldn't build Router Service: couldn't instantiate query planner; invalid schema: schema validation errors: There can be only one directive named "@defer".
```

We probably want to only append the `@defer` directive if it's not part of the supergraph schema yet.

**Comment by o0Ignition0o** [comment]
The [built in directives are here](https://github.com/apollographql/federation/blob/b3a3cb84d8d67d1d6e817dc85b9ae0ecdd9908d1/internals-js/src/definitions.ts#L1131) and I think [this function](https://github.com/apollographql/federation/blob/b3a3cb84d8d67d1d6e817dc85b9ae0ecdd9908d1/internals-js/src/directiveAndTypeSpecification.ts#L60) is where we (already) check for presence before we add them.

STRS:

run the router against this supergraph (that specified defer and stream in one of its subgraphs)

```graphql
schema
  @core(feature: "https://specs.apollo.dev/core/v0.1")
  @core(feature: "https://specs.apollo.dev/join/v0.1") {
  query: Query
}

directive @core(feature: String!) repeatable on SCHEMA

directive @join__field(
  graph: join__Graph
  requires: join__FieldSet
  provides: join__FieldSet
) on FIELD_DEFINITION

directive @join__type(
  graph: join__Graph!
  key: join__FieldSet
) repeatable on OBJECT | INTERFACE

directive @join__owner(graph: join__Graph!) on OBJECT | INTERFACE

directive @join__graph(name: String!, url: String!) on ENUM_VALUE

scalar join__FieldSet

enum join__Graph {
  MYSERVICE @join__graph(name: "myservice", url: "http://somewhere")
}

"""
The `@defer` directive may be provided for fragment spreads and inline fragments to inform the executor to delay the execution of the current fragment to indicate deprioritization of the current fragment. A query with `@defer` directive will cause the request to potentially return multiple responses, where non-deferred data is delivered in the initial response and data deferred is delivered in a subsequent response. `@include` and `@skip` take precedence over `@defer`.
"""
directive @defer(
  """
  If this argument label has a value other than null, it will be passed on to the result of this defer directive. This label is intended to give client applications a way to identify to which fragment a deferred result belongs to.
  """
  label: String
  """
  Deferred when true.
  """
  if: Boolean
) on FRAGMENT_SPREAD | INLINE_FRAGMENT

"""
The `@stream` directive may be provided for a field of `List` type so that the backend can leverage technology such as asynchronous iterators to provide a partial list in the initial response, and additional list items in subsequent responses. `@include` and `@skip` take precedence over `@stream`.
"""
directive @stream(
  """
  If this argument label has a value other than null, it will be passed on to the result of this stream directive. This label is intended to give client applications a way to identify to which fragment a streamed result belongs to.
  """
  label: String
  """
  The initial elements that shall be send down to the consumer.
  """
  initialCount: Int! = 0
  """
  Streamed when true.
  """
  if: Boolean
) on FIELD

type Query {
  hello: String!
}
```

Expected: the router spins up
Actual:

```
2022-10-19T11:47:38.622427Z ERROR cannot create the router: couldn't build Router Service: couldn't instantiate query planner; invalid schema: schema validation errors: There can be only one directive named "@defer".
caused by

2022-10-19T11:47:38.622792Z ERROR could not create the HTTP pipeline: couldn't build Router Service: couldn't instantiate query planner; invalid schema: schema validation errors: There can be only one directive named "@defer".
caused by

could not create the HTTP pipeline: couldn't build Router Service: couldn't instantiate query planner; invalid schema: schema validation errors: There can be only one directive named "@defer".
caused by
```

**Comment by o0Ignition0o** [comment]
I think the discrepancy happens because the tool that was used [to generate the schema](https://github.com/ChilliCream/hotchocolate/blob/80350d82253419a9ff631bbbaae0f640eb834753/src/HotChocolate/Core/src/Types/Types/Directives/DeferDirective.cs#L15) made the defer condition `Boolean` while we expect `Boolean!` that defaults to `true`

```javascript
args: [
        { name: 'label', type: schema.stringType() },
        { name: 'if', type: new NonNullType(schema.booleanType()), defaultValue: true },
      ],
```

Which should have translated in graphql to:

```graphql
if: Boolean! = true
```

I m not quite sure how to go about this, so I'll ask other teams how to reconcile this best.

**Comment by abernix** [comment]
@pcmanus I'm wondering if you have any thoughts here.  I'm curious if the Router should just unconditionally add its own definition of `@defer` to the supergraph though I'm hesitant to generally transform the supergraph itself in ways that aren't observable in the actual supergraph artifact.  Perhaps it's actually more that `toAPISchema` should produce that?

I'm unsure also unsure what `@defer` in the supergraph would do in the case of passing that same supergraph artifact to the Gateway considering the Gateway doesn't support `@defer`.  In one school of thought, I'm tempted to say that the _same_ Supergraph artifact should be able to be passed to the Gateway or the Router — sans `@defer` — and the Router should itself take on the burden of allowing clients to use that though maybe it should only do that via a flag to `toAPISchema` in the Router's invocation of that (but perhaps not in Studio infrastructure) or something?

Sorry a bit of a ramble, but Thoughts?

**Comment by o0Ignition0o** [comment]
> I believe the 2nd point is already what we do, but unfortunately it doesn't seem we do the 1st point correctly and that may be why we encounter issues at the moment when subgraphs do define @defer

I think so too, AIUI [checkOrAdd](https://github.com/apollographql/federation/blob/b3a3cb84d8d67d1d6e817dc85b9ae0ecdd9908d1/internals-js/src/directiveAndTypeSpecification.ts#L60) checks for the full signature, and not just the name.

One way to go about it would be to turn `checkOrAdd` into a `replace` that removes a directive by name and inserts ours, but I don't know what this would imply for other built in directives.

An other possibility would be to have `builtInDirectives` that only get added if they're not defined, and an other set of directives that would always overwrite whatever has been already defined.

**Comment by pcmanus** [comment]
> One way to go about it would be to turn checkOrAdd into a replace that removes a directive by name and inserts ours, but I don't know what this would imply for other built in directives.

Not sure we're thinking of the same thing. I was mostly referring the [`toGraphQLJSchema`](https://github.com/apollographql/federation/blob/main/internals-js/src/definitions.ts#L1300), which afaict, is the method that is used by the router to ensure the defer definition is part of the API schema, or at least it used to be.

If there is a need to change the `checkOrAdd` method, then I think we're not aligned on what the problem is as again, I don't think we should be manually adding the defer definition to the supergraph schema at all.

**Comment by pcmanus** [comment]
Fyi, I pushed a https://github.com/apollographql/federation/pull/2212 with what I think fix the problem currently seen with the router failing when there is some `@defer` definition in the supergraph.
